build_tools/pexpect_helper.py

```diff
@@ -0,0 +1,261 @@
+"""pexpect_helper provides a wrapper around the pexpect module.
+
+This module exposes a single class SpawnedProc, which wraps pexpect.spawn().
+This exposes a pseudo-tty, which fish or another process may talk to.
+The send() function may be used to send data to fish, and the expect_* family
+of functions may be used to match what is output to the tty.
+
+Example usage:
+  sp = SpawnedProc() # this launches fish
+  sp.expect_prompt() # wait for a prompt
+  sp.sendline("echo hello world")
+  sp.expect_prompt("hello world")
+
+"""
+
+from __future__ import print_function
+
+import inspect
+import os
+import os.path
+import re
+import sys
+import time
+import pexpect
+
+# Default timeout for failing to match.
+TIMEOUT_SECS = 5
+
+
+def get_prompt_re(counter):
+    """ Return a regular expression for matching a with a given prompt counter. """
+    return re.compile(
+        r"""(?:\r\n?|^)   # beginning of line
+            (?:\[.\]\ )?  # optional vi mode prompt
+         """
+        + (r"prompt\ %d>" % counter),  # prompt with counter
+        re.VERBOSE,
+    )
+
+
+def get_callsite():
+    """ Return a triple (filename, line_number, line_text) of the call site location. """
+    callstack = inspect.getouterframes(inspect.currentframe())
+    for f in callstack:
+        if inspect.getmodule(f.frame) is not Message.MODULE:
+            return (os.path.basename(f.filename), f.lineno, f.code_context)
+    return ("Unknown", -1, "")
+
+
+def escape(s):
+    """ Escape the string 's' to make it human-understandable. """
+    res = []
+    for c in s:
+        if c == "\n":
+            res.append("\\n")
+        elif c == "\r":
+            res.append("\\r")
+        elif c == "\t":
+            res.append("\\t")
+        elif c.isprintable():
+            res.append(c)
+        else:
+            res.append("\\x{:02x}".format(ord(c)))
+    return "".join(res)
+
+
+class Message(object):
+    """ Some text either sent-to or received-from the spawned proc.
+
+    Attributes:
+        dir: the message direction, either DIR_SEND or DIR_RECV
+        filename: the name of the file from which the message was sent
+        text: the text of the messages
+        when: a timestamp of when the message was sent
+    """
+
+    DIR_SEND = "SENT"
+    DIR_RECV = "RECV"
+    MODULE = sys.modules[__name__]
+
+    def __init__(self, dir, text, when):
+        """ Construct from a direction, message text and timestamp. """
+        self.dir = dir
+        self.filename, self.lineno, _ = get_callsite()
+        self.text = text
+        self.when = when
+
+    @staticmethod
+    def sent(text, when):
+        """ Return a SEND message with the given text. """
+        return Message(Message.DIR_SEND, text, when)
+
+    @staticmethod
+    def received(text, when):
+        """ Return a RECV message with the given text. """
+        return Message(Message.DIR_RECV, text, when)
+
+    def formatted(self):
```

Comment:
I'm not entirely up on idiomatic python. Should this not be `__str__`? Then you could remove the explicit call later.

Reply:
So I've done some work on this, and I've found it nicer to remove the formatting responsibility from this class, so you can do relative timestamps.